File: src/main/java/com/github/sblundy/elvish/psi/ElvishParameterBase.java
package com.github.sblundy.elvish.psi;

import com.intellij.extapi.psi.ASTWrapperPsiElement;
import com.intellij.lang.ASTNode;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiNameIdentifierOwner;
import com.intellij.util.IncorrectOperationException;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public abstract class ElvishParameterBase extends ASTWrapperPsiElement implements PsiNameIdentifierOwner {
    public ElvishParameterBase(@NotNull ASTNode node) {
        super(node);
    }

    @Nullable
    @Override
    public PsiElement getNameIdentifier() {
        return getCompound();
    }

    @Override
    public PsiElement setName(@NotNull String name) throws IncorrectOperationException {
        //TODO is actually a variable name?
        ElvishVariableName ne = ElvishPsiUtils.INSTANCE.newNameElement(name, getProject());
        getCompound().replace(ne);
        return this;
    }

    @NotNull
    public abstract ElvishCompound getCompound();
}

File: src/main/kotlin/com/github/sblundy/elvish/psi/ElvishParameterBase.kt
package com.github.sblundy.elvish.psi

import com.github.sblundy.elvish.psi.ElvishPsiUtils.newNameElement
import com.intellij.icons.AllIcons
import com.intellij.lang.ASTNode
import com.intellij.navigation.ItemPresentation
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiNameIdentifierOwner
import com.intellij.util.IncorrectOperationException
import javax.swing.Icon

abstract class ElvishParameterBase(node: ASTNode) : ASTWrapperElvishPsiElement(node), PsiNameIdentifierOwner, ElvishParameter {
    override fun getNameIdentifier(): PsiElement? = getVariableName()

    override fun getName(): String? = getVariableName().text

    @Throws(IncorrectOperationException::class)
    override fun setName(name: String): PsiElement {
        //TODO is actually a variable name?
        val ne = newNameElement(name, project)
        getVariableName().replace(ne)
        return this
    }

    override fun getIcon(flags: Int): Icon? = AllIcons.Nodes.Parameter

    override fun getPresentation(): ItemPresentation? = ElvishBasicItemPresentation(getVariableName().text, AllIcons.Nodes.Parameter)
}